crossbar/worker/proxy.py

```diff
@@ -0,0 +1,537 @@
+        self._authid = kw.get('authid', 'anonymous')
+        self._authrole = kw.get('authrole', None)
+        self._args = kw
+
+    @property
+    def authextra(self):
+        return self._args.get(u'authextra', dict())
+
+    def on_challenge(self, session, challenge):
+        raise RuntimeError("on_challenge called on proxy authentication")
+
+    def on_welcome(self, msg, authextra):
+        self.log.info(
+            "AuthProxy.on_welcome: {message} {extra}",
+            message=msg,
+            extra=authextra,
+        )
+        return {
+            "authextra": {
+                "authid": self._authid,
+                "realm": self._realm,
+            }
+        }
+
+
+auth.IAuthenticator.register(AuthProxy)
+
+
+class ProxyServerSession(RouterSession):
+    """
+    The autobahn.wamp.interfaces.ITransportHandler for Proxies. This
+    forwards every incoming message to the other side's transport
+    (with no further processing) with the *exception* of the messages
+    in .do_not_forward
+    """
+    _backend = None
+    log = txaio.make_logger()
+
+    # types of messages we do not forward but pass up to the parent
+    # class for normal-style Router processing
+    do_not_forward = (
+        message.Hello,
+        message.Welcome,
+        message.Authenticate,
+    )
+
+    def set_backend(self, other_session):
+        self._backend = other_session
+
+    def onMessage(self, msg):
+        if isinstance(msg, self.do_not_forward):
+            return super(ProxyServerSession, self).onMessage(msg)
+        # TODO: should always destroy 'the other side' if one side
+        # (including the backend) goes away (e.g. message IDs would
+        # get out of sync if we just auto re-connected to the backend)
+        if not self._backend or not self._backend._transport:
+            raise RuntimeError("No backend, or backend transport gone")
+        self._backend._transport.send(msg)
+
+
+class ProxyRealm(object):
+    """
+    Faking out some things so we can provide a valid _realm object for
+    ProxyRouter
+    """
+
+    def __init__(self, realm_id, session):
+        self._realm_id = realm_id
+        self.session = session  # the authenticator (among others?) will call on this session
+
+
+class ProxyRouter(object):
+    """
+    Provide enough Router functionality so that crossbar's base
+    RouterSession class can call various methods on us..
+    """
+    log = txaio.make_logger()
+
+    def __init__(self, realm, worker, transport_config):
+        self._realm = ProxyRealm(realm, worker)
+        self._worker = worker
+        self._front_to_back = dict()  # session_id -> ProxySession backend
+        self._backend_transport_config = transport_config
+
+    def has_role(self, role):
+        """
+        we just forward any role at all onto the 'real' backend
+
+        XXX TODO: is this true? or should we config which roles are
+        okay? or is that up to 'the real backend' itself?
+        """
```

Comment:
This probably does want to be a configuration-thing (and/or a hook-function a user can pass in somewhere).